File: easy/1380_lucky_numbers_in_a_matrix/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<int> luckyNumbers(std::vector<std::vector<int>> &matrix) {
    const int rows = matrix.size();
    const int cols = matrix[0].size();

    std::vector<int> luckyNums;

    for (std::vector<int> &row : matrix) {
      int rowMin = *min_element(row.begin(), row.end());
      int col = distance(row.begin(), find(row.begin(), row.end(), rowMin));

      int colMax = 0;
      for (int r = 0; r < rows; ++r)
        colMax = std::max(matrix[r][col], colMax);

      if (rowMin == colMax)
        luckyNums.push_back(rowMin);
    }

    return luckyNums;
  }
};

File: easy/1380_lucky_numbers_in_a_matrix/solution.py
from typing import List


class Solution:
    def luckyNumbers(self, matrix: List[List[int]]) -> List[int]:
        rows = len(matrix)
        cols = len(matrix[0])

        min_num_by_row = [0] * rows
        max_num_by_col = [0] * cols

        for row in range(rows):
            min_num_by_row[row] = min(matrix[row])

        for col in range(cols):
            max_num_by_col[col] = max(matrix[row][col] for row in range(rows))

        lucky_nums = []

        for row in range(rows):
            for col in range(cols):
                if min_num_by_row[row] == max_num_by_col[col]:
                    lucky_nums.append(matrix[row][col])

        return lucky_nums
